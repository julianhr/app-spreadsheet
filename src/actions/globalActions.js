import { createAction } from 'redux-starter-kit'


export const setActiveCell = createAction('SET_ACTIVE_CELL')
export const displayInputter = createAction('DISPLAY_INPUTTER')
