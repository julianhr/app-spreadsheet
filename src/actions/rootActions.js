import { createAction } from 'redux-starter-kit'


export const setActiveCell = createAction('SET_ACTIVE_CELL')

export function moveActiveCell(key) {
  return (dispatch, getState) => {
    const { activeCell, rows, columns } = getState()
    const currRow = activeCell.codePointAt(0) - 'a'.codePointAt(0)
    const currCol = Number.parseInt(activeCell[1]) - 1
    let nextRow = currRow
    let nextCol = currCol

    if (key === 'ArrowUp') { nextRow = currRow - 1 }
    else if (key === 'ArrowRight') { nextCol = currCol + 1 }
    else if (key === 'ArrowDown') { nextRow = currRow + 1 }
    else if (key === 'ArrowLeft') { nextCol = currCol - 1 }

    if (Math.min(nextRow, nextCol) >= 0 && nextRow < rows && nextCol < columns) {
      nextRow = String.fromCodePoint('a'.codePointAt(0) + nextRow)
      nextCol = nextCol + 1
      dispatch(setActiveCell(`${nextRow}${nextCol}`))
    }
  }
}
