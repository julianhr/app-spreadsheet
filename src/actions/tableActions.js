import graph from '~/formulas/graph'
import Interpreter from '../formulas/Interpreter'


export function setCellData(location, entered) {
  const interpreter = new Interpreter(location)
  const result = interpreter.interpret(entered)


  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function replaceCellData(location, entered, result) {
  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function clearCellData(location) {
  return (dispatch, getState) => {
    const value = getState().table[location]

    graph.delVertex(location)

    if (value === undefined) {
      return Promise.resolve()
    }


    return Promise.resolve(dispatch({
      type: 'DELETE_CELL_DATA',
      payload: location,
    }))
  }
}
