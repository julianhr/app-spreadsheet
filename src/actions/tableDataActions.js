export function setCellData(location, entered, result) {
  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function clearCellData(location) {
  return (dispatch, getState) => {
    const value = getState().tableData[location]

    if (value === undefined) {
      return Promise.resolve()
    }

    return Promise.resolve(dispatch({
      type: 'DELETE_CELL_DATA',
      payload: location,
    }))
  }
}
