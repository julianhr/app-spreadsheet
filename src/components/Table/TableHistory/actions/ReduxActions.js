import reduxConnect from '../reduxConnect'


class ReduxActions {
  static reduxConnect = reduxConnect

  get reduxConnect() {
    return ReduxActions.reduxConnect
  }

  getCellData(location) {
    return this.reduxConnect.getState().tableData[location]
  }
}

export default ReduxActions
