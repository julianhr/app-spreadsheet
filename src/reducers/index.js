import { configureStore } from 'redux-starter-kit'

import globalReducer from './globalReducer'
import tableReducer from './tableReducer'


const appStoreFn = () => (
  configureStore({
    reducer: {
      global: globalReducer,
      table: tableReducer,
    }
  })
)


export default appStoreFn
