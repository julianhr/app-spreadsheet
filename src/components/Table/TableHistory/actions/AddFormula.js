import ReduxActions from './ReduxActions'


class AddFormula extends ReduxActions {
  constructor(location, oldVertex, newVertex) {
    super()
    this.location = location
    this.entered = newVertex.entered
    this.result = newVertex.result
    this.previous = oldVertex && {
      entered: oldVertex.entered,
      result: oldVertex.result,
    }
  }

  execute() {
    this.reduxConnect.fn.setCellData(this.location, this.entered, this.result)
    this.reduxConnect.fn.setActiveCell(this.location)
  }

  unexecute() {
    if (this.previous) {
      const { entered, result } = this.previous
      this.reduxConnect.fn.setCellData(this.location, entered, result)
    } else {
      this.reduxConnect.fn.clearCellData(this.location)
    }

    this.reduxConnect.fn.setActiveCell(this.location)
  }
}

export default AddFormula
