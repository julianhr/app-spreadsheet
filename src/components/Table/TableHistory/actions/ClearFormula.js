import ReduxActions from './ReduxActions'


class ClearFormula extends ReduxActions {
  constructor(location) {
    super()
    this.previous = this.getCellData(location)
    this.location = location
  }

  execute() {
    this.reduxConnect.fn.clearCellData(this.location)
    this.reduxConnect.fn.setActiveCell(this.location)
  }

  unexecute() {
    if (!this.previous) { return }
    const { entered, result } = this.previous
    this.reduxConnect.fn.setCellData(this.location, entered, result)
    this.reduxConnect.fn.setActiveCell(this.location)
  }
}

export default ClearFormula