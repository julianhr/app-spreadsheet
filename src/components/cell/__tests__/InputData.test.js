import React from 'react'
import { create } from 'react-test-renderer'
import { render, fireEvent, cleanup } from '@testing-library/react'

import ConnectedInputData, { InputData } from '../InputData'
import MockApp from '~/__tests__/__mocks__/MockApp'
import { appStoreGen } from '~/reducers'
import { setCellData } from '~/actions/tableDataActions'
import {
  ERR_DIVISION_BY_ZERO,
  ERR_CIRCULAR_REFERENCE,
  ERR_GENERIC,
} from '~/formulas/Interpreter'


const requiredProps = {
  // props
  replaceValue: false,
  location: 'B-2',
  onEscape: jest.fn(),
  onCommit: jest.fn(),
  // redux
  clearCellData: jest.fn(),
  setCellData: jest.fn(),
}

const testProps = {
  ...requiredProps,
  // props
  entered: 'test entered',
}

const createApp = (props) => {
  jest.spyOn(InputData.prototype, 'componentDidMount').mockReturnValueOnce('mocked cdm')
  const wrapper = create(
    <MockApp>
      <InputData {...props} />
    </MockApp>
  )
  const { instance } = wrapper.root.findByType(InputData)
  return [wrapper, instance]
}

const renderApp = (props, customStore) => {
  const store = customStore || appStoreGen()
  const wrapper = render(
    <MockApp customStore={store}>
      <ConnectedInputData {...props} />
    </MockApp>
  )
  return [wrapper, store]
}

describe('InputData', () => {
  afterEach(cleanup)

  describe('props', () => {
    it('does not raise warning if al required props are passed', () => {
      expect(() => createApp(requiredProps)).not.toThrow()
    })

    it('raises warning if required props is not present', () => {
      Object.keys(requiredProps).forEach(key => {
        const props = {...testProps}
        delete props[key]
        expect(() => createApp(props)).toThrow()
      })
    })
  })

  describe('#focusInputTag', () => {
    it('focuses input tag when component mounts', () => {
      const [wrapper, _] = renderApp(testProps)
      const input = wrapper.getByDisplayValue('')
      expect(document.activeElement).toBe(input)
    })

    it('places cursor on far-right of formula string', async () => {
      const mockRef = { focus: jest.fn(), scrollLeft: 0, scrollWidth: 25 }
      const [_, instance] = createApp(testProps)
      instance.refInput.current = mockRef
  
      instance.focusInputTag()
      expect(instance.refInput.current.focus).toHaveBeenCalledTimes(1)
      expect(instance.refInput.current.scrollLeft).toEqual(mockRef.scrollWidth)
    })
  })

  describe('#setNewValue', () => {
    it('if string is empty the current value is deleted', async () => {
      const appStore = appStoreGen()
      const location = testProps.location
      const isEnteredValid = true
      let entered, result

      entered = 'test string'
      result = entered
      await appStore.dispatch(setCellData(location, entered, isEnteredValid, result))
      expect(appStore.getState().tableData[location]).toBeTruthy()

      renderApp(testProps, appStore)
      const input = document.querySelector(`[data-location="${location}"]`)
      input.value = ''
      fireEvent.keyDown(input, { key: 'Enter' })
      expect(appStore.getState().tableData[location]).toBeUndefined()
    })
  })

  describe('#handleOnKeyDown', () => {
    test('sequence of events if key is "Escape"', () => {
      testProps.onEscape.mockReset()
      renderApp(testProps)
      const input = document.querySelector('[data-cell="input"]')
      fireEvent.keyDown(input, { key: 'Escape' })
      expect(testProps.onEscape).toHaveBeenCalledTimes(1)
    })

    test('sequence of event if key is "Enter', () => {
      testProps.onCommit.mockReset()
      renderApp(testProps)
      const input = document.querySelector('[data-cell="input"]')
      fireEvent.keyDown(input, { key: 'Enter' })
      expect(testProps.onCommit).toHaveBeenCalledTimes(1)
    })

    describe('bubbling', () => {
      describe('Function suggestions not displayed', () => {
        test('only if key is "Enter" or "Escape"', () => {
          const events = 'b Enter 2 , t Escape E s 5'.split(' ')
          const spyKeyDown = jest.fn()
          const Listener = ({ children }) => <div onKeyDown={spyKeyDown}>{children}</div>
          render(
            <MockApp>
              <Listener>
                <InputData {...testProps}  />
              </Listener>
            </MockApp>
          )
          const input = document.querySelector('[data-cell="input"]')
    
          events.forEach(key => fireEvent.keyDown(input, { key }))
          expect(spyKeyDown).toHaveBeenCalledTimes(2)
        })
      })
    })

    describe('Function suggestions is displayed', () => {
      xtest('only if key is "Enter" or "Escape"', () => {
        const events = 'b Enter 2 , t Escape E s 5'.split(' ')
        const spyKeyDown = jest.fn()
        const Listener = ({ children }) => <div onKeyDown={spyKeyDown}>{children}</div>
        render(
          <MockApp>
            <Listener>
              <InputData {...testProps}  />
            </Listener>
          </MockApp>
        )
        const input = document.querySelector('[data-cell="input"]')
        fireEvent.keyDown(input, { key: '=' })
        fireEvent.keyDown(input, { key: 's' })
  
        events.forEach(key => fireEvent.keyDown(input, { key }))
        expect(spyKeyDown).toHaveBeenCalledTimes(0)
      })
    })
  })

  describe('#handleOnBlur', () => {
    test('stores value and calls #onCommit', () => {
      const [_, store] = renderApp(testProps)
      const entered = 'test value'
      const cellData = { entered, result: entered }

      testProps.onCommit.mockReset()
      const input = document.querySelector('[data-cell="input"]')
      input.value = entered
      fireEvent.blur(input)

      expect(store.getState().tableData[testProps.location]).toEqual(cellData)
      expect(testProps.onCommit).toHaveBeenCalledTimes(1)
    })
  })

  describe('functional tests', () => {
    const funcProps = {
      // props
      replaceValue: true,
      location: 'B-2',
      onEscape: jest.fn(),
      onCommit: jest.fn(),
    }

    it('displays default value if replaceValue is false', async () => {
      const { location } = testProps
      const entered = 'test entered'
      const result = 'test result'
      const store = appStoreGen()
      await store.dispatch(setCellData(location, entered, true, result))
      const props = {...funcProps}
      props.replaceValue = false
      renderApp(props, store)
      const el = document.querySelector(`[data-location="${location}"]`)

      expect(el).toMatchSnapshot()
    })

    it('will not display default value if replaceValue is true', async () => {
      const { location } = testProps
      const entered = 'test entered'
      const result = 'test result'
      const store = appStoreGen()
      await store.dispatch(setCellData(location, entered, true, result))
      const props = {...funcProps}
      props.replaceValue = true
      renderApp(props, store)
      const el = document.querySelector(`[data-location="${location}"]`)

      expect(el).toMatchSnapshot()
    })
  })
})
