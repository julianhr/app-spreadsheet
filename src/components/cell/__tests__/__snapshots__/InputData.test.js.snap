// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputData functional tests displays default value if replaceValue is false 1`] = `
<DocumentFragment>
  <input
    class="css-te4i1p-Input egik0ru0"
    id="B-2"
    value="test formula"
  />
</DocumentFragment>
`;

exports[`InputData functional tests will not display default value if replaceValue is true 1`] = `
<DocumentFragment>
  <input
    class="css-te4i1p-Input egik0ru0"
    id="B-2"
    value=""
  />
</DocumentFragment>
`;
